### PR TITLE
Allow all Cross-Origin Methods

### PR DESCRIPTION
**Pull request by blink1073**
Allows all REST API methods to be cross-origin, if explicitly requested with `--NotebookApp.allow_origin="*"`

### CODE REVIEW

**Comment by minrk** [comment]
This is explicitly reintroducing loads of XSS vulnerabilities, which I don't think we want.

**Comment by blink1073** [comment]
Even though it is only turned available when `--NotebookApp.allow_origin="*"`?

**Comment by minrk** [comment]
Ah, then maybe I misunderstand. If it only has an effect when allow_origin is already whitelisting, doesn't this do nothing at all?

**Comment by blink1073** [comment]
When you send a cross-origin PATCH or DELETE request, it also sends an OPTION request.  This allows that transaction to take place.

**Comment by blink1073** [comment]
I tested it on both this patch and on Kernel DELETE.

**Comment by minrk** [comment]
Arg, I need to do more testing. I hate CORS more than just about anything.

If you want this to affect all API handlers, it should go on the base APIHandler class, rather than on every APIHandler subclass.

**Comment by blink1073** [comment]
But not all of them support the same methods

**Comment by minrk** [comment]
But CORS restrictions shouldn't be relevant to that. Is there an 'allow-methods: *'?

**Comment by blink1073** [comment]
Sure, we could do that and then explicitly add OPTIONS to all of the SUPPORTED_METHODS

**Comment by minrk** [comment]
By the way, I just checked and I can make cross-origin API requests with current master without this PR if I do `--allow-origin='*'`, so I'm not sure what this accomplishes.

**Comment by blink1073** [comment]
With PATCH and DELETE methods?

**Comment by blink1073** [comment]
It looks like `Access-Control-Allow-Methods` must be a comma separated, capitalized list.

**Comment by minrk** [comment]
Aha! Now I understand, sorry. Default Access-Control-Allow-Methods is just get and post (of course, what I tried with). Adding all the verbs to APIHandler.options makes perfect sense. I understand now that this PR only makes the server behave exactly how I thought it already behaved.

**Comment by blink1073** [comment]
So, add a full set of `options` to `APIHandler`, and then OPTIONS to all of the SUPPORTED_METHODS?

**Comment by minrk** [comment]
@blink1073 yup. I don't think any of the current `options` handlers need to exist, and a single `options` handler on the base APIHandler should be the right thing.

**Comment by minrk** [comment]
@blink1073 I think you can actually delete the `SUPPORTED_METHODS` attribute on all of the handlers, but I'll check on that.

**Comment by blink1073** [comment]
I explicitly needed to add it to make it work.  We can add it to `APIHandler` instead.

**Comment by minrk** [comment]
Yup, instead of adding it, you can remove SUPPORTED_METHODS from all the subclasses. The default value from RequestHandler should be fine.

**Comment by blink1073** [comment]
Okay, overhaul coming right up.

File: notebook/services/api/handlers.py [hunk 1]
```diff
@@ -18,6 +18,13 @@ def get(self):
         self.set_header('Content-Type', 'text/x-yaml')
         return web.StaticFileHandler.get(self, 'api.yaml')
 
+    @web.authenticated
+    def options(self, section_name):
```
Comment (minrk) [inline notebook/services/api/handlers.py]: I think you will need to accept `*args, **kwargs`  to deal with tornado's URL parsing

Comment (blink1073) [inline notebook/services/api/handlers.py]: Fixed, but I'm still getting rejected on OPTIONS requests.

**Comment by blink1073** [comment]
Weird, the method doesn't get called if it is in the base class.  What kind of shenanigans is tornado pulling here?

File: notebook/services/api/handlers.py [hunk 2]
```diff
@@ -18,6 +18,13 @@ def get(self):
         self.set_header('Content-Type', 'text/x-yaml')
         return web.StaticFileHandler.get(self, 'api.yaml')
 
+    @web.authenticated
+    def options(self, *args, **kwargs):
```
Comment (minrk) [inline notebook/services/api/handlers.py]: ah, sorry for the confusion. This is just a super-bad choice of names. This isn't be base APIHandler class, this is a weird, only partially implemented handler for providing the REST API spec. The base class is in base/handlers.py

**Comment by blink1073** [comment]
Nevermind, I'm just dumb as usual...

**Comment by blink1073** [comment]
Okay, this now passes my test suite that has DELETE and PATCH method calls.

**Comment by blink1073** [comment]
Do you ever sleep @minrk?

**Comment by minrk** [comment]
:+1: I'd love a test for it, but I think testing cross-origin behavior should be a separate project.

**Comment by minrk** [comment]
(and yes, I happen to be in bed right now!)

**Comment by blink1073** [comment]
good, you go to bed I'll go get dinner.  Thanks again, Min.

**Comment by episodeyang** [comment]
@blink1073 thanks for fixing this! I just ran into this problem [here](https://github.com/ipython/ipython/issues/8849)

**Comment by zischwartz** [comment]
The weird thing about this is the lack of `OPTIONS` should have broken Thebe (which always runs cross origin) long ago, but I'm only seeing it now.

**Comment by episodeyang** [comment]
I think the only api method that needs this would be to halt the kernel. This is used in switching kernel types!

**Comment by rgbkrk** [comment]
The API call that @zischwartz ran into with the new version of the notebook + thebe (built from 4.x js) was the initial creation of the kernel.

**Comment by episodeyang** [comment]
I see! @zischwartz what API call is that? (method and end point). I just want to know so that I can keep it in mind

**Comment by minrk** [comment]
Initial creation is POST, so should have been allowed without this patch.

**Pull request merged**